### PR TITLE
Fix overlaping condition

### PR DESCRIPTION
**Pull request by Thomasb81**
Fix proposition for https://github.com/antlr/antlr4/issues/2547

### CODE REVIEW

**Comment by ericvergnaud** [comment]
@lygav what do you think?

**Comment by Thomasb81** [comment]
If it can ease the review, here are the  equivalent line in respectively the java runtime
https://github.com/antlr/antlr4/blob/7c334b114c20b5d0718fb5db7334c3ca0c462fe0/runtime/Java/src/org/antlr/v4/runtime/TokenStreamRewriter.java#L521

and the Cpp runtime.

https://github.com/antlr/antlr4/blob/7c334b114c20b5d0718fb5db7334c3ca0c462fe0/runtime/Cpp/runtime/src/TokenStreamRewriter.cpp#L361

It look like historically the java code also test `!disjoin && !same ` :
https://github.com/antlr/antlr4/blob/bf9c7c3a28c04e0f8ad2e304416d828d308894ec/runtime/Java/src/org/antlr/v4/runtime/TokenStreamRewriter.java#L489
but has been optimize later...

**Comment by lygav** [comment]
@ericvergnaud Long time since i wrote that, cant say i have an opinion. Quickly reviewing the linked issue and this PR i think it's fine.

**Comment by ericvergnaud** [comment]
@parrt blessed

**Comment by ericvergnaud** [comment]
@lygav thanks, I didn't write that code, so just wanted to make sure everybody is comfortable

**Comment by ericvergnaud** [comment]
> If it can ease the review, here are the equivalent line in respectively the java runtime
> 
> https://github.com/antlr/antlr4/blob/7c334b114c20b5d0718fb5db7334c3ca0c462fe0/runtime/Java/src/org/antlr/v4/runtime/TokenStreamRewriter.java#L521
> 
> and the Cpp runtime.
> 
> https://github.com/antlr/antlr4/blob/7c334b114c20b5d0718fb5db7334c3ca0c462fe0/runtime/Cpp/runtime/src/TokenStreamRewriter.cpp#L361
> 
> It look like historically the java code also test `!disjoint && !same ` :
> 
> https://github.com/antlr/antlr4/blob/bf9c7c3a28c04e0f8ad2e304416d828d308894ec/runtime/Java/src/org/antlr/v4/runtime/TokenStreamRewriter.java#L489
> 
> 
> but has been optimize later...

@mike-lischke  seems the cpp runtime also requires fixing?

**Comment by mike-lischke** [comment]
Yes, if the Java version no longer checks `same` then it makes sense to update the C++ runtime as well. @Thomasb81 could you add this change to your patch? Would avoid me having to create a super mini PR exclusively for that. Thanks.

**Comment by ericvergnaud** [comment]
Seems CSharp version also needs the same fix

**Comment by Thomasb81** [comment]
It look like my troubles to use git are now exposed ...

Commit to fix the cpp, python and Csharp runtime...

File: runtime/Cpp/runtime/src/TokenStreamRewriter.cpp [hunk 1]
```diff
@@ -358,7 +357,7 @@ std::unordered_map<size_t, TokenStreamRewriter::RewriteOperation*> TokenStreamRe
         rop->lastIndex = std::max(prevRop->lastIndex, rop->lastIndex);
         std::cout << "new rop " << rop << std::endl;
       }
-      else if (!disjoint && !same) {
+      else if ( !disjoint ) {
```
Comment (mike-lischke) [inline runtime/Cpp/runtime/src/TokenStreamRewriter.cpp]: Please honor the existing formatting style. No spaces around expressions in parentheses.

File: runtime/Python2/src/antlr4/TokenStreamRewriter.py [hunk 2]
```diff
@@ -157,13 +157,12 @@ def _reduceToSingleOperationPerIndex(self, rewrites):
...
                     continue
                 isDisjoint = any((prevRop.last_index < rop.index, prevRop.index > rop.last_index))
-                isSame = all((prevRop.index == rop.index, prevRop.last_index == rop.last_index))
                 if all((prevRop.text is None, rop.text is None, not isDisjoint)):
                     rewrites[prevRop.instructionIndex] = None
                     rop.index = min(prevRop.index, rop.index)
                     rop.last_index = min(prevRop.last_index, rop.last_index)
                     print('New rop {}'.format(rop))
-                elif (not(isDisjoint) and not(isSame)):
+                elif ( not(isDisjoint) ):
```
Comment (mike-lischke) [inline runtime/Python2/src/antlr4/TokenStreamRewriter.py]: same here

File: runtime/Python3/src/antlr4/TokenStreamRewriter.py [hunk 3]
```diff
@@ -156,13 +156,12 @@ def _reduceToSingleOperationPerIndex(self, rewrites):
...
                     continue
                 isDisjoint = any((prevRop.last_index<rop.index, prevRop.index>rop.last_index))
-                isSame = all((prevRop.index == rop.index, prevRop.last_index == rop.last_index))
                 if all((prevRop.text is None, rop.text is None, not isDisjoint)):
                     rewrites[prevRop.instructionIndex] = None
                     rop.index = min(prevRop.index, rop.index)
                     rop.last_index = min(prevRop.last_index, rop.last_index)
                     print('New rop {}'.format(rop))
-                elif (not(isDisjoint) and not(isSame)):
+                elif ( not(isDisjoint) ):
```
Comment (mike-lischke) [inline runtime/Python3/src/antlr4/TokenStreamRewriter.py]: and here

**Comment by Thomasb81** [comment]
It look like the travis-ci failure is unrelated  to the content of my PR ...

**Comment by mike-lischke** [comment]
I restarted the build and the tests succeed, but there's a Travis error:

> home/travis/.travis/functions: line 524:  6055 Terminated              travis_jigger "${!}" "${timeout}" "${cmd[@]}"
> cat: target/surefire-reports/*.dumpstream: No such file or directory

I guess @parrt needs to look after that.

**Comment by parrt** [comment]
It looks like Travis is just having some problems. Merging. thanks everyone!

**Pull request merged**